added support for parsing hash on vector elements

KageKirin opened the pull request:
Hello,

this is a WIP pull request for issue "Attributes for types in vector" #4498.
It is incomplete for the moment, i.e. the initial commit in this PR adds the parsing support,
but the parsed values are not applied inside the code generators yet.

I will update this PR as the changes progress.

Regards.

KageKirin commented:
added so far:
- parser support for `fields : [ulong](hash: "fnv1a", cpp_type: "FieldT");`
- C++ code generation of above definition as `std::vector<FieldT*> fields;`
- C++ code generation of initializer of above field as `: fields()`

missing yet:
- rehasher generation
- resolver generation

aardappel commented:
Can you please run `generate_code.sh` and add changed files to the PR, so we can see what the effect of the codegen is. You may need to add a field that is a vector of hashes to the schema.

aardappel commented on src/idl_parser.cpp:
```diff
@@ -633,7 +633,7 @@ CheckedError Parser::ParseField(StructDef &struct_def) {
   auto hash_name = field->attributes.Lookup("hash");
   if (hash_name) {
-    switch (type.base_type) {
+    switch (IsScalar(type.base_type) ? type.base_type : type.element) {
```
should we check here that if it is not scalar, it is actually a vector?

KageKirin replied:
fair point.
I reversed the logic to check for BASE_TYPE_VECTOR instead of IsScalar.

KageKirin commented on src/idl_gen_cpp.cpp:
```diff
@@ -1304,6 +1308,11 @@ class CppGenerator : public BaseGenerator {
                   Name(field) + "(" + native_default->constant + ")";
             }
           }
+        } else if(field.value.type.base_type == BASE_TYPE_VECTOR) {
```
Actually, I don't think this part is needed, as vectors are initialised just fine without a call to their empty ctor.
Also, this makes a class' initializer list longer, which is not intended.

KageKirin commented:
Follow up: 
The `Pack`-ing part of this works in my small test case, provided a custom `rehasher_function_t`.
The `UnPack`-ing part of this works as well, given a custom `resolver_function_t`.

So, I think the functionality can be merged now.
What do you think?

Regards.

KageKirin commented:
Oh, I see merge conflicts. Do you want me to rebase the changes first and then update this pull request?

aardappel commented:
Yes, please rebase.

KageKirin commented:
Ok, so I rebased and tested the code on my little example from #4613 as well as on my full codebase.
Looks good so far.

aardappel commented on src/idl_gen_cpp.cpp:
```diff
@@ -446,7 +446,7 @@ class CppGenerator : public BaseGenerator {
   // Return a C++ type from the table in idl.h
   std::string GenTypeBasic(const Type &type, bool user_facing_type) const {
     static const char *ctypename[] = {
-    // clang-format off
```
why were these spaces removed?

KageKirin replied:
oh! looks like a merge error. will fix this in the next rebase.

aardappel commented on src/idl_gen_cpp.cpp:
```diff
@@ -1645,7 +1649,9 @@ class CppGenerator : public BaseGenerator {
 
         code_ += "  const {{CPP_NAME}} *{{FIELD_NAME}}_nested_root() const {";
-        code_ += "    return flatbuffers::GetRoot<{{CPP_NAME}}>({{FIELD_NAME}}()->Data());";
+        code_ +=
```
string on the same line?

aardappel commented:
sorry, looks like another rebase is needed.

KageKirin commented:
no problem.
I'll also condense the changelists into fewer ones if you're fine with it.

KageKirin commented:
Alright.
I rebased from master, condensed the commits into fewer and fixed the whitespace issues.

Would be great to have this merged pretty soon: I have another WIP with a fix for #4615 that I'd like to pull-request this week.

Cheers.

aardappel commented:
Ok, Thanks!

KageKirin commented:
You're welcome!

aardappel commented:
@amitsaini12345 just mentioned on https://gitter.im/google/flatbuffers that this PR broke the VS C# build (missing Referrable.cs etc).
